source/unifiedbuffer.ts

```diff
@@ -0,0 +1,142 @@
+
+/* spellchecker: disable */
+
+import { Buffer } from './buffer';
+import { Context } from './context';
+import { Initializable } from './initializable';
+
+/* spellchecker: enable */
+
+
+export class UnifiedBuffer extends Initializable {
+
+    protected _cpuBuffer: ArrayBuffer;
+    protected _gpuBuffer: Buffer;
+    protected _updates = new Array<Update>();
+
+    usage: GLenum;
+
+    constructor(context: Context, sizeInBytes: number, usage: GLenum, identifier?: string) {
+        super();
+
+        this._cpuBuffer = new ArrayBuffer(sizeInBytes);
+        this._gpuBuffer = new Buffer(context, identifier);
+        this.usage = usage;
+    }
+
+    protected _addUpdate(update: Update): void {
+        const toRemove = new Array<number>();
+        const toMerge = new Array<Update>();
+
+        // Mark all older updates which overlap with the new update
+        // for removing / merging
+        this._updates.forEach((current: Update, index: number) => {
+            if (!(update.end < current.begin || update.begin > current.end)) {
+                toRemove.push(index);
+                toMerge.push(current);
+            }
+        });
+
+        // Remove all older updates, since they get merged in the next step
+        for (let i = toRemove.length - 1; i >= 0; i--) {
+            this._updates.splice(toRemove[i], 1);
+        }
+
+        // Finally merge all overlapping updates
+        toMerge.push(update);
```

Comment:
I suggest to make this strategy parameterizable:
* Introduce a threshold parameter (unit: bytes)
* This threshold is used to to collapse ranges whose distance is below the threshold
* A threshold value of `-1` indicates that all ranges should be collapsed

Reply:
Should this threshold apply only to new ranges or to all existing ones as well?

If it should only apply to the new range, do we collaps only the 0-2 closest ranges with the new range or expand from there on and collaps all ranges iteratively, which are inside the new collapsed range?
Example: |---Left old---|     |---Right old---|     |---New---|
Assuming the ranges are _merge threshold_ bytes from each other, do we collaps only the New and Right old ranges or do we collaps all ranges, since Left old is inside the merge threshold after New and Right old got collapsed.

Reply:
I suggest to collapse all ranges, transitively.